### ISSUE TITLE
Error in tt-mlir uplift

### ISSUE THREAD

**Issue by vmilosevic** [tenstorrent/tt-forge-fe#309]
Running tests with latest tt-mlir fails with the following errors:

```
2024-09-18 09:06:21.033 | FATAL    | TTDevice        - Tensor 0 - data type mismatch: expected UInt8, got UInt32

2024-09-18 09:06:28.814 | FATAL    | TTDevice        - Tensor 1 - data type mismatch: expected Float32, got BFloat16

2024-09-18 09:09:45.831 | INFO     | forge.compiled_graph_state:__call__:330 - Running model DualConv2d on device...
                 Always | FATAL    | Unsupported data type
```
                 
For a full log look at the CI run https://github.com/tenstorrent/tt-forge-fe/actions/runs/10918549924
PR https://github.com/tenstorrent/tt-forge-fe/pull/307

**Comment by vmilosevic** [comment]
Another issue is that after the last error, the test run goes into a loop instead of exiting
```
terminate called recursively
tt_forge_signal_handler - signal: 6 (abort)
terminate called recursively
tt_forge_signal_handler - signal: 6 (abort)
terminate called recursively
tt_forge_signal_handler - signal: 6 (abort)
terminate called recursively
```

**Comment by nvukobratTT** [comment]
Looks like dual conv2d fails during runtime:

```
...
2024-09-18 09:09:45.831 | INFO     | forge.compiled_graph_state:__call__:330 - Running model DualConv2d on device...
                 Always | FATAL    | Unsupported data type
terminate called after throwing an instance of 'std::runtime_error'
                  Metal | INFO     | Closing device 0
                  Metal | INFO     | Closing device 0
  what():  TT_THROW @ /__w/tt-forge-fe/tt-forge-fe/third_party/tt-mlir/third_party/tt-metal/src/tt-metal/ttnn/cpp/ttnn/tensor/tensor_impl.cpp:42: tt::exception
info:
Unsupported data type
backtrace:
 --- tt::tt_metal::tensor_impl::element_size_bytes(tt::tt_metal::DataType)
 --- tt::tt_metal::create_device_tensor(tt::tt_metal::Shape const&, tt::tt_metal::DataType, tt::tt_metal::Layout, tt::tt_metal::Device*, tt::tt_metal::MemoryConfig const&)
...
```

@pilkicTT do you happen to have some context around this failure? It fails on one of the conv tests. @vmilosevic did the pipeline fail on the same test after the re-run?

**Comment by vmilosevic** [comment]
Yes, this is reproducible. @dgolubovicTT also noticed the same issue

**Comment by pilkicTT** [comment]
The following commit: https://github.com/tenstorrent/tt-mlir/commit/2e170337a092a3bff05494b3b7b16a375261daf4 is triggering the issue.

We should probably revert it...

**Comment by nvukobratTT** [comment]
> Yes, this is reproducible. @dgolubovicTT also noticed the same issue

@vmilosevic can you open up an issue for @jnie-TT to disable this by default? 

@nsmithtt @jnie-TT are there any functional benefits at the moment that we're utilizing with async mode? Do we have enough context how and when it works on tt-meta?

**Comment by vmilosevic** [comment]
@nvukobratTT Sorry, I don't have enough context on what needs to change and why. It's best if the component owner can create this in a ticket. 
I would like to see some basic RCA on this, what was broken and why, and how we can cover this with tests.

**Comment by nsmithtt** [comment]
I think @jnie-TT  has a real fix for this, that said, @jnie-TT let's disable async mode for now, we can reenable at some point in the future, but it's not needed for V0.

**Comment by nvukobratTT** [comment]
> I think we should also establish a policy:
> - tt-mlir uplifts metal, gets green PR
> - Does _not_ land
> - tt-forge-fe uplifts tt-mlir and points to PR branch, gets green
> - Only after green forge-fe can we land tt-mlir
> - forge-fe uplifts at its convenience 
> 
> I think we should do this until we have the mono-repo setup.

Yup, this makes sense. 

To add on this, FFE vs. MLIR should be uplifted more often. To do so, we can have 2 proces in place:
1. FFE + MLIR integration testing. If FFE tests are passing, we create PR automatically and merge changes.
2. FFE + MLIR + Metal. What Nick currently proposed. 

If we see value in more often Metal uplifts, we can merge the two into single process.

**Comment by nvukobratTT** [comment]
Discussion continued on PR as well:
- https://github.com/tenstorrent/tt-forge-fe/pull/307

**Comment by nvukobratTT** [comment]
@vmilosevic we managed to find the root cause for a good chunk of mentioned failures. Here is the first issue on the MLIR side that describes which change causes FFE regressions:
- https://github.com/tenstorrent/tt-mlir/issues/813

As @sdjordjevicTT isn't here, can you @mtopalovicTT take a look into this regression, or find someone who can check this change out? :))

**Comment by nvukobratTT** [comment]
@sdjordjevicTT seems like the latest MLIR solves this uplift. I didn't have time to check commit diff, but one of the latest changes solved existing issues.

In sum, there is potential that we just hid the bug. For now, we're going to uplift MLIR as all tests are passing in order to unblock Llama 3B, but if you have time please double-check what happened and confirm is this fix intentional, and does it solves issue at hand :)) Once you confirm it, feel free to close this issue :))

**Comment by sdjordjevicTT** [comment]
@svuckovicTT will look at it next week when he comes back from vacation.

**Comment by JushBJJ** [comment]
Having the same issue as well with this simple code:
```python
import forge
import torch

from torch import nn

class foo(nn.Module):
    def __init__(self):
        super().__init__()
        self.weights = nn.Parameter(torch.randn(1024, 2048, dtype=torch.bfloat16))
    
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return torch.nn.functional.linear(x, self.weights)

def test_bug():
    model = foo()
    inputs = torch.rand((1, 12, 2048), dtype=torch.bfloat16)
    golden_output = model(inputs) # passes
    forge_model = forge.compile(model, sample_inputs=inputs)

    out = forge_model(inputs)
    print(out)

test_bug()
```

Error:
```
2024-10-06 10:16:29.431 | FATAL    | TTDevice        - Tensor 1 - data type mismatch: expected Float32, got BFloat16
Traceback (most recent call last):
  File "/projects/TT-Forge/bug/the_bug.py", line 23, in <module>
    test_bug()
  File "/projects/TT-Forge/bug/the_bug.py", line 20, in test_bug
    out = forge_model(inputs)
  File "/tt-forge-fe/forge/forge/compiled_graph_state.py", line 218, in __call__
    model_outputs = run_binary(self.compiled_binary, int(ProgramId.FORWARD), inputs_and_parameters)
RuntimeError: Tensor 1 - data type mismatch: expected Float32, got BFloat16
```

It looks like it stems from somewhere in the TVM repo.
`tt-forge-fe/third_party/tvm/python/tvm/contrib/forge_compile.py`:
![image](https://github.com/user-attachments/assets/c27d8085-b9db-4ab6-9ce6-d7a11bc88852)

`tt-forge-fe/third_party/tvm/python/tvm/contrib/forge_utils.py`:
![image](https://github.com/user-attachments/assets/209e7d6c-db33-4e32-95af-bce7447c4a81)

**Comment by JushBJJ** [comment]
This is a super weird bug, im having this issue too with this simple code:
```python
import forge
import torch

from torch import nn

class foo(nn.Module):
    def __init__(self):
        super().__init__()
        self.weights = nn.Parameter(torch.randn(1024, 2048, dtype=torch.bfloat16))
    
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return torch.nn.functional.linear(x, self.weights)

def test_bug():
    model = foo()
    inputs = torch.rand((1, 12, 2048), dtype=torch.bfloat16)
    golden_output = model(inputs) # passes
    forge_model = forge.compile(model, sample_inputs=inputs)

    out = forge_model(inputs)
    print(out)

test_bug()
```

The issue relies somewhere in the tvm repo it looks like, it stems when `with ConvertEmulatedDtypes(torchmod, inputs)` is called in `tt-forge-fe/third_party/tvm/python/tvm/contrib/forge_compile.py` in the `compile_pytorch_for_forge` function.

![2024-10-06-204410_1004x734_scrot](https://github.com/user-attachments/assets/52c4dfb2-6adb-436a-9929-72a3241e9bc0)

It converts self.inputs' dtype from bfloat16 to the fallback float32 dtype

**Comment by JushBJJ** [comment]
Even if we remove bfloat16 from self.emulated_dfs, the compiled model now expects float32 even if inputs stays as bfloat16 🤷‍♂️
![2024-10-06-210236_1411x791_scrot](https://github.com/user-attachments/assets/89dd5e6c-fa1f-4b7c-ad38-279be2088f79)

**Comment by nvukobratTT** [comment]
@sdjordjevicTT @svuckovicTT can we create a followup issue for this one on the MLIR repo? Just to close it on FFE as it's not a blocker anymore :))

**Comment by nvukobratTT** [comment]
> I'm finding it a bit hard to follow what's going on in this issue... Is there an outstanding problem that needs to be dealt with, and is it on the forge or MLIR side? @nvukobratTT

This isn't a blocker anymore for FFE.

This commit was causing issues for a specific group of tests, however one of the newer ones at that time solved the issue. However, we're not sure if we just hid the underlying bug or if it was solved. All in all, feel free to sync with @sdjordjevicTT regarding your priorities, and is this something worth exploring atm :))
